crates/core/src/metadata_store/providers/objstore/immutable_store.rs

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+
+use bytes::Bytes;
+use bytestring::ByteString;
+
+use crate::metadata_store::{Precondition, VersionedValue, WriteError};
+use restate_types::errors::GenericError;
+use restate_types::Version;
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
+pub(crate) struct ImmutableStore {
+    /// a version that is increased with every mutating operation.
+    /// do not confuse with the per-key versions.
+    global_version: Version,
+    /// The collection of versioned values.
+    /// in the context of the metadata store, this is expected to be very small,
+    /// and the kv entries are also cheap to clone (Arc)
+    map: BTreeMap<ByteString, VersionedValue>,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, Clone)]
+#[serde(tag = "version", content = "value")]
+enum OnDiskImmutableStore<'a> {
+    V1(Cow<'a, ImmutableStore>),
+}
+
+impl Default for ImmutableStore {
+    fn default() -> Self {
+        Self {
+            global_version: Version::MIN,
+            map: Default::default(),
+        }
+    }
+}
+
+/// This is the core logic
+impl ImmutableStore {
+    pub(crate) fn serialize(&self) -> Result<Bytes, GenericError> {
+        let on_disk = OnDiskImmutableStore::V1(Cow::Borrowed(self));
+
+        let mut buf = vec![];
+
+        ciborium::into_writer(&on_disk, &mut buf)
+            .map(|_| Bytes::from(buf))
+            .map_err(Into::into)
+    }
+
+    pub(crate) fn deserialize(bytes: impl AsRef<[u8]>) -> Result<Self, GenericError> {
+        let on_disk: OnDiskImmutableStore<'static> = ciborium::from_reader(bytes.as_ref())?;
+
+        match on_disk {
+            OnDiskImmutableStore::V1(cow) => Ok(cow.into_owned()),
+        }
+    }
+
+    pub(crate) fn current_version(&self) -> Version {
+        self.global_version
+    }
+
+    pub(crate) fn update_map(
+        &self,
+        update_fn: impl FnOnce(&mut BTreeMap<ByteString, VersionedValue>),
+    ) -> Self {
+        let mut new_map = self.map.clone();
+        update_fn(&mut new_map);
+
+        Self {
+            global_version: self.global_version.next(),
```

Comment:
I'm wondering if approaching this as a `into_builder()` / `build_next_version()` might not be a safer API. Chaining multiple `update_map` calls will create gaps in the version sequence, which we want to avoid.